Deprecate Application Support and Update Migration Guides

AliSoftware opened the pull request:
* [x] I've started my branch from the `develop` branch (gitflow)
  * [x] And I've selected `develop` as the "base" branch for this Pull Request I'm about to create
* [x] I've added an entry in the `CHANGELOG.md` file to explain my changes and credit myself  
      _(or added `#trivial` to the PR title if I think it doesn't warrant a mention in the CHANGELOG)_
  * [x] Add the entry in the appropriate section (Bug Fix / New Feature / Breaking Change / Internal Change)
  * [x] Add a period and 2 spaces at the end of your short entry description
  * [x] Add links to your GitHub profile to credit yourself and to the related PR and issue after your description

---

SwiftGen-Eve commented:
<!--
  0 Errors
  0 Warnings
  0 Messages
  5 Markdowns
-->

Hey 👋 I'm Eve, the friendly bot watching over SwiftGen 🤖

Thanks a lot for your contribution!



---

Seems like everything is in order 👍 You did a good job here! 🤝

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AliSoftware commented on CHANGELOG.md:
```diff
@@ -120,7 +122,7 @@
   [David Jennes](https://github.com/djbe)
   [#613](https://github.com/SwiftGen/SwiftGen/pull/613)
-* Update to SwiftLint 0.31.0 and enable some extra SwiftLint rules.  
+* Update SwiftLint and enable some extra SwiftLint rules.  
```
Not sure when but we later updated to SwiftLint 0.39.2, so I figured we shouldn't let people think when reading the CHANGELOG that we're still on 0.31

djbe replied:
Good catch!

djbe commented on Documentation/MigrationGuide.md:
```diff
@@ -11,12 +11,39 @@ All the migration guides for SwiftGen are spread out over a few files, depending
+
+## Deprecation of Application Support lookup for named templates
+
+In previous versions of SwiftGen, SwiftGen looked for named templates inside `~/Library/Application Support` before looking for bundled templates. This feature was little known and complexified both code and documentation, in addition to making your project dependent on what was installed on your machine. This is why it has been is deprecated in SwiftGen 6.2 and will be removed in 7.0.
```
Complexity of code doesn't matter to the user. But the docs were confusing, so maybe:
```suggestion
In previous versions of SwiftGen, SwiftGen looked for named templates inside `~/Library/Application Support` before looking for bundled templates. This feature was little known and made the documentation quite confusing, in addition to making your project dependent on what was installed on your machine. This is why it has been is deprecated in SwiftGen 6.2 and will be removed in 7.0.
```

djbe commented on Documentation/MigrationGuide.md:
```diff
@@ -11,12 +11,39 @@ All the migration guides for SwiftGen are spread out over a few files, depending
+
+The invocation of individual parsers via the command line (`swiftgen <parser>`, e.g. `swiftgen colors …`, `swiftgen strings …`) is deprecated in favor of using the `swiftgen run <parser>` subcommand instead.
+
+The `swiftgen <parser>` command will be completely removed in SwiftGen 7.0, only leaving `swiftgen run <parser>` available
```
```suggestion
The `swiftgen <parser>` command will be completely removed in SwiftGen 7.0, only leaving the `swiftgen run <parser>` subcommand.
```

djbe commented on Documentation/MigrationGuide.md:
```diff
@@ -11,12 +11,39 @@ All the migration guides for SwiftGen are spread out over a few files, depending
+
+#### `swiftgen templates` renamed `swiftgen template`
+
+The command to manage templates (list bundled templates, print their content to generate custom variations, open documentation…) has been renamed `swiftgen template` (singular).
```
```suggestion
The command to manage templates (list bundled templates, print their content to generate custom templates, open documentation…) has been renamed `swiftgen template` (singular).
```

djbe commented on Documentation/MigrationGuide.md:
```diff
@@ -11,12 +11,39 @@ All the migration guides for SwiftGen are spread out over a few files, depending
+
+The command to manage templates (list bundled templates, print their content to generate custom variations, open documentation…) has been renamed `swiftgen template` (singular).
+
+`swiftgen templates` (plural) is thus deprecated and will be removed in SwiftGen 7.0
```
```suggestion
`swiftgen templates` (plural) is thus deprecated and will be removed in SwiftGen 7.0.
```

djbe commented on Documentation/templates/MigrationGuide.md:
```diff
@@ -1,4 +1,27 @@
+
+### Removed Swift 3 templates
+
+We've removed the bundled `swift3` templates, as we added `swift5` templates. We assume most of you are already developing in a newer version of Swift (4 or 5). If not, you can copy the `swift3` templates from older versions of SwiftGen into your project and point to them using `templatePath`.
```
```suggestion
We've removed the bundled `swift3` templates, as we added `swift5` templates. We assume most of you are already developing in a newer version of Swift (4 or 5), as recent Xcode versions no longer provide support for Swift 3. If you're still using an old Swift version, you can copy the `swift3` templates from older versions of SwiftGen into your project and point to them using `templatePath`.
```

djbe commented on Documentation/templates/MigrationGuide.md:
```diff
@@ -1,4 +1,27 @@
+
+### CoreData: entityName is now a property
+
+If you used Core Data templates, the `entityName` is now correctly a `class var` property instead of a `class func`. You will have to remplace calls to `.entityName()` by `.entityName` instead.
```
```suggestion
If you are using Core Data templates, `entityName` is now correctly a `class var` property instead of a `class func`. You will have to remplace calls to `.entityName()` by `.entityName` instead.
```

djbe commented on Documentation/templates/MigrationGuide.md:
```diff
@@ -1,4 +1,27 @@
+
+### Strings: `%@` now generates an `Any` instead of `String`
+
+When generating constants for strings, if your string contains `%@` placeholders, we previously generated a function with corresponding parameters of type `String`. We now generate these parameters as being of type `Any`, since `%@` corresponds to any class or `NSObject`, not just `String`.
```
```suggestion
When generating constants for strings, if your string contains `%@` placeholders, we previously generated a function with corresponding parameters of type `String`. We now correctly generate these parameters as being of type `Any`, since `%@` corresponds to any class or `NSObject`, not just `String`.
```

djbe commented on Documentation/templates/MigrationGuide.md:
```diff
@@ -1,4 +1,27 @@
+
+When generating constants for strings, if your string contains `%@` placeholders, we previously generated a function with corresponding parameters of type `String`. We now generate these parameters as being of type `Any`, since `%@` corresponds to any class or `NSObject`, not just `String`.
+
+This means that if you were previously explicitly calling `obj.description` or `String(describing: obj)` to be able to pass an object `obj` as a `String` to one of the generated strings functions for localization, you can now replace your code to directly pass `obj` instead.
```
```suggestion
This means that if you were previously explicitly calling `obj.description` or `String(describing: obj)` to be able to pass an object `obj` to one of the generated strings functions for localization, you can now replace your code to directly pass `obj` instead.
```

djbe commented on Documentation/templates/MigrationGuide.md:
```diff
@@ -1,4 +1,27 @@
+
+### XCAsset templates typealiases deprecation
+
+In the unlikely case that you used the `colorAliasName` and `imageAliasName` typealiases generated by the `XCAsset` templates, know that they are now deprecated. You should get a proper depreciation notice by Xcode if you use them anyway.
```
```suggestion
In the unlikely case that you used the `colorAliasName` and `imageAliasName` typealiases generated by the `XCAsset` templates, know that they are now deprecated and will be removed in SwiftGen 7.0. You should get a proper depreciation notice by Xcode if you use them anyway.
```

djbe commented on Sources/SwiftGen/TemplateRef.swift:
```diff
@@ -45,8 +45,18 @@ enum TemplateRef: Equatable {
+        logMessage(
+          .warning,
+          """
+          Referring templates in Application Support by name will be deprecated in SwiftGen 7.0.
```
```suggestion
          Referring to templates in Application Support by name will be deprecated in SwiftGen 7.0.
```

AliSoftware replied:
```suggestion
          Referring to templates in Application Support by name is deprecated and will be removed in SwiftGen 7.0.
```

djbe replied:
Eh yeah, that 🤦

AliSoftware commented on Documentation/MigrationGuide.md:
```diff
@@ -11,12 +11,39 @@ All the migration guides for SwiftGen are spread out over a few files, depending
+
+The invocation of individual parsers via the command line (`swiftgen <parser>`, e.g. `swiftgen colors …`, `swiftgen strings …`) is deprecated in favor of using the `swiftgen run <parser>` subcommand instead.
+
+The `swiftgen <parser>` command will be completely removed in SwiftGen 7.0, only leaving the `swiftgen run <parser>` subcommand.
```
```suggestion
The `swiftgen <parser>` commands will be completely removed in SwiftGen 7.0, only leaving the `swiftgen run <parser>` subcommands.
```

AliSoftware commented:
@djbe addressed feedbacks + fixed conflict after we merged the other PR about custom L10n function. If you could formerly approve the latest changes we should then be able to merge now 🙂
